### PR TITLE
OAuth

### LINKED ISSUE #36: [frontend/backend] Add oauth logic for Github/Twitter/Reddit
Don't connect to database, just go through the flow and callback. Database connection/identity registration will be in a different task.

### LINKED ISSUE #38: [frontend] change signup flow UI/UX to oauth with Github/Twitter/Reddit.
after designer has designed the new flow and UI

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #36 
Closes #38 

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by vimwitch** [comment]
Test UI available [here](http://test.unirep.io:49154/).

@kittybest @AnthonyMadia can we use  this pr instead of #93 and #59

**Comment by AnthonyMadia** [comment]
The signup flow is pretty clean. I signed up with twitter and made a post. Tried signing up again with Twitter and got the json response `You have already signed up with this account'` which is expected behavior. 

I have no issue using this PR. I was trying to figure the best way to programmatically login with Twitter for Cypress e2e test  following the GoogleOauth Cypress example in their [docs](https://docs.cypress.io/guides/end-to-end-testing/google-authentication#:~:text=used%20with%20testing.-,Setting%20Google%20app%20credentials%20in%20Cypress,-To%20have%20access) but was having issues with this.

**Comment by vimwitch** [comment]
Yea I'm not really sure about testing. In the past I've just not written tests and been really careful with oauth logic, manually testing changes.

**Comment by AnthonyMadia** [comment]
Even with a custom command like: 
```
Cypress.Commands.add('loginByGoogleApi', () => {
  cy.log('Logging in to Google')
  cy.request({
    method: 'POST',
    url: 'https://www.googleapis.com/oauth2/v4/token',
    body: {
      grant_type: 'refresh_token',
      client_id: Cypress.env('googleClientId'),
      client_secret: Cypress.env('googleClientSecret'),
      refresh_token: Cypress.env('googleRefreshToken'),
    },
  }).then(({ body }) => {
    const { access_token, id_token } = body

    cy.request({
      method: 'GET',
      url: 'https://www.googleapis.com/oauth2/v3/userinfo',
      headers: { Authorization: `Bearer ${access_token}` },
    }).then(({ body }) => {
      cy.log(body)
      const userItem = {
        token: id_token,
        user: {
          googleId: body.sub,
          email: body.email,
          givenName: body.given_name,
          familyName: body.family_name,
          imageUrl: body.picture,
        },
      }

      window.localStorage.setItem('googleCypress', JSON.stringify(userItem))
      cy.visit('/')
    })
  })
})
```
I just was not sure about the exact parameters that would make the Twitter Auth work. It seems possible though once we have the correct env variables, endpoints, and body in the requests

**Comment by vimwitch** [comment]
I was thinking we add a flag in the frontend enabling an automatic signup flow during tests. e.g. pressing a signup button causes instant signup and skips the oauth redirects.

**Comment by AnthonyMadia** [comment]
Sounds good

**Comment by ChialiTsai** [comment]
reddit is not functional?

**Comment by ChialiTsai** [comment]
image
using github, can we add a loader here? it feels a bit dead here...there was about 2 sec of wait.

**Comment by kittybest** [comment]
> I was thinking we add a flag in the frontend enabling an automatic signup flow during tests. e.g. pressing a signup button causes instant signup and skips the oauth redirects.

So we don't have to run `test:e2e` including testing the oauth things?

**Review by kittybest — APPROVED** [review]

**Comment by vimwitch** [comment]
This is almost ready. Just adding the sign in functionality and an e2e test doing `sign up -> sign out -> sign in`.

@ChialiTsai What do you think of only including Twitter and Github oauth?

File: packages/frontend/cypress/support/commands.ts [hunk 1]
```diff
@@ -47,26 +48,46 @@ Cypress.Commands.add('deployUnirep', () => {
                     transaction: tx.hash,
                 })
             })
```
Comment (AnthonyMadia) [inline packages/frontend/cypress/support/commands.ts:50]: Add an alias here

File: packages/frontend/cypress/support/commands.ts [hunk 2]
```diff
@@ -47,26 +48,46 @@ Cypress.Commands.add('deployUnirep', () => {
...
+            cy.intercept(`${serverUrl}/api/oauth/twitter?*`, {
+                statusCode: 301,
+                headers: {
+                    Location:
+                        'http://localhost:3000/start?signupCode=test_signup_code',
+                },
+            })
```
Comment (AnthonyMadia) [inline packages/frontend/cypress/support/commands.ts:57]: Add an alias here

**Review by AnthonyMadia — CHANGES_REQUESTED** [review]

**Comment by AnthonyMadia** [comment]
Is there any reason for getting this error locally: `Error: anondb error: NotFoundError: Failed to execute 'index' on 'IDBObjectStore': The specified index was not found.` as an uncaught exception error?

**Comment by vimwitch** [comment]
Not sure about that, trying running `yarn` in the root directory, i bumped the `anondb` version.

**Review by AnthonyMadia — APPROVED** [review]

**Comment by AnthonyMadia** [comment]
yea not sure why I get this error locally. Ran yarn in root and tried anondb@0.0.12 and anondb@0.0.13 giving same error with e2e tests. If this is just happening on my machine I don't want to hold up the PR.

**Comment by kittybest** [comment]
@ChialiTsai  I also agree with only use github and twitter to sign up, since reddit is less used, and it has different logic with twitter and github (which are all using oauth2.0 now)

**Comment by kittybest** [comment]
For this PR, I found some using bugs:
1. if I type in wrong password, still redirect to homepage without any clue that the password is wrong, although I actually logged in failed.
2. the password can be empty, maybe when it's empty, then disable the encrypt button.

but I think these are bugs and feature adjustments, might be another issues.

**Review by kittybest — APPROVED** [review]

**Comment by vimwitch** [comment]
> if I type in wrong password, still redirect to homepage without any clue that the password is wrong, although I actually logged in failed.
> the password can be empty, maybe when it's empty, then disable the encrypt button.

Both fixed.

I think reddit isn't very good for sybil resistance. Github and Twitter are marginally better.

**Comment by ChialiTsai** [comment]
> > if I type in wrong password, still redirect to homepage without any clue that the password is wrong, although I actually logged in failed.
> > the password can be empty, maybe when it's empty, then disable the encrypt button.
> 
> Both fixed.
> 
> I think reddit isn't very good for sybil resistance. Github and Twitter are marginally better.

ok got it~
I was thinking some of design thinking for Unirep social is very much related to reddit, therefore I thought it was useful to have it so we can migrate user to here. but I guess for this wave, the user will most likely have github.

**Pull request merged**